app/models/user_access.rb

```diff
@@ -0,0 +1,82 @@
+class UserAccess < Struct.new(:user)
+  class NotAuthorizedError < StandardError; end
+
+  LEVELS = {
+    viewer: {
+      id: :viewer,
+      name: "View: Everything",
+      grant_access: [],
+      description: "Can view stuff"
+    },
+
+    manager: {
+      id: :manager,
+      name: "Manager",
+      grant_access: [:viewer, :manager],
+      description: "Can manage stuff"
+    },
+
+    power_user: {
+      id: :power_user,
+      name: "Power User",
+      description: "Can manage everything"
+    }
+  }.freeze
+
+  ACTION_TO_LEVEL = {
+    view: [:manager, :viewer],
+    manage: [:manager]
+  }.freeze
+
+  def accessible_organizations(action)
+    resources_for(Organization, action)
+  end
+
+  def accessible_facility_groups(action)
+    resources_for(FacilityGroup, action)
+      .or(FacilityGroup.where(organization: accessible_organizations(action)))
+  end
+
+  def accessible_facilities(action)
+    resources_for(Facility, action)
+      .or(Facility.where(facility_group: accessible_facility_groups(action)))
+  end
+
+  def can?(action, model, record = nil)
+    if record&.is_a? ActiveRecord::Relation
+      raise ArgumentError, "record should not be an ActiveRecord::Relation."
+    end
+
+    case model
+      when :facility
+        can_access_record?(accessible_facilities(action), record)
+      when :organization
+        can_access_record?(accessible_organizations(action), record)
+      when :facility_group
+        can_access_record?(accessible_facility_groups(action), record)
+      else
+        raise ArgumentError, "Access to #{model} is unsupported."
+    end
+  end
+
+  private
+
+  def can_access_record?(resources, record)
+    return true if user.power_user?
+    return resources.find_by_id(record).present? if record
+    resources.exists?
+  end
+
+  def resources_for(resource_model, action)
+    return resource_model.all if user.power_user?
+    return resource_model.none unless ACTION_TO_LEVEL.fetch(action).include?(user.access_level.to_sym)
```

Comment:
This clause has been moved up, it used to be like [this](https://github.com/simpledotorg/simple-server/pull/1242/files#diff-dbee6fa62dba5eb76fea1e205e2a5f54L58) before. It makes more sense this way since we're just pulling accesses off a concrete `User`: